feat: free the blobs

g11tech opened the pull request:
Free the blobs

the branch is now functional! (cleanup and some blob gossip validations still wip)

Below are the logs of a local devnet (with a corresponding ethereumjs branch) where peer1 is producing blobs
And peer2 syncs block and blobs
1. Initially via req/resp to get `Synced`
2. Then via gossip to stay `Synced`
![image](https://user-images.githubusercontent.com/76567250/224494375-af94db14-b531-45f7-8d08-f76ef11a5142.png)
3. Unknown Parent root sync using beacon/blob side cars by root
![image](https://user-images.githubusercontent.com/76567250/225609010-67a093a3-0a04-4a6a-9f27-f1e070b0d575.png)



ref:
- https://github.com/ethereum/consensus-specs/pull/3244

TODOS: 
 - [x] modified block/blob production
 - [x] modified blob dbs
 - [x] handled block/blob gossip 
 - [x] marry blob and block gossip for BlockInput
 - [x] updated references    
 - [x] fixed gossip types
 - [x] updated fetch blocksMaybeBlobsbyRange
 - [x] updated fetch blocksMaybeBlobsbyRoot
 - [x] handle block/blob archival to finalizeddbs    
 - [x] blob publish
 - [x] fix block publish
 - [x] handle onBlobSideCarByRange
 - [x] handle onBlockSideCarbyRoot
 - [x] prune gossip blockinput cache
 - [x] handle genesis state blobs db handling
 - [x] validator blob signing and publishing
   - [x] modify get produced blobs
   - [x] sign and propagate
 - [x] integrate c-kzg with computeBlobKzgProof
   - [x]  update interfaces


TODO:
- [x] single node run with ethereumjs courtsey: - https://github.com/ethereumjs/ethereumjs-monorepo/pull/2567
- [x] Multi peer run to test gosssip
- [x] Multi peer run to sync to test req/resp
  - [x] normal sync via blobssidecar by range
  - [x] unknown block sync via by root
- [x]  Fix beaconBlockMayBeBlobsByRoot request: review and make it multi block

wemeetagain commented on packages/beacon-node/src/chain/blocks/types.ts:
```diff
@@ -11,7 +11,7 @@ export enum BlockInputType {
 export type BlockInput =
   | {type: BlockInputType.preDeneb; block: allForks.SignedBeaconBlock}
-  | {type: BlockInputType.postDeneb; block: allForks.SignedBeaconBlock; blobs: deneb.BlobsSidecar};
+  | {type: BlockInputType.postDeneb; block: allForks.SignedBeaconBlock; blobs: deneb.BlobsSidecar[]};
```
Should this be BlobSidecar[] ?

Also a few other places where BlobsSidecar remains

g11tech replied:
wip :slightly_smiling_face:

dapplion commented on packages/beacon-node/src/network/reqresp/handlers/blobSidecarsByRoot.ts:
```diff
@@ -0,0 +1,42 @@
+  for (const blobIdentifier of requestBody) {
+    const {blockRoot} = blobIdentifier;
+    const blockRootHex = toHex(blockRoot);
+    const summary = chain.forkChoice.getBlockHex(blockRootHex);
```
Since there any no other block types in this context just name this `block`

dapplion commented on packages/beacon-node/src/network/reqresp/handlers/blobSidecarsByRoot.ts:
```diff
@@ -0,0 +1,42 @@
+    // SPEC: Clients MUST support requesting blocks and sidecars since the latest finalized epoch.
+    // https://github.com/ethereum/consensus-specs/blob/11a037fd9227e29ee809c9397b09f8cc3383a8c0/specs/eip4844/p2p-interface.md#beaconblockandblobssidecarbyroot-v1
+    if (!summary || summary.slot <= finalizedSlot) {
+      // TODO: Should accept the finalized block? Is the finalized block in the archive DB or hot DB?
```
No, the spec clearly states to not return requests for finalized

dapplion commented on packages/beacon-node/src/network/reqresp/beaconBlocksMaybeBlobsByRoot.ts:
```diff
@@ -12,72 +12,21 @@ export async function beaconBlocksMaybeBlobsByRoot(
+  const blobSidecars = await Promise.all(Array.from({length: blobKzgCommitmentsLen}),(_v,index)=>{
+    const blobRequest = [{blockRoot,index}]
+    return reqResp.blobSidecarsByRootRequest(peerId,blobRequest)
+  })
```
This function is a WIP right?

g11tech replied:
finished :+1:

dapplion commented on packages/beacon-node/src/network/network.ts:
```diff
@@ -514,13 +509,14 @@ export class Network implements INetwork {
       {type: GossipType.voluntary_exit},
       {type: GossipType.proposer_slashing},
       {type: GossipType.attester_slashing},
+      {type: GossipType.beacon_block},
```
Note the commented line at 507

dapplion commented on packages/beacon-node/src/network/gossip/handlers/index.ts:
```diff
@@ -121,6 +121,41 @@ export function getGossipHandlers(modules: ValidatorFnsModules, options: GossipH
     }
   }
 
+  async function validateBlobSidecar(
```
If this function is used once, just inline below

dapplion commented on packages/beacon-node/src/network/gossip/gossipsub.ts:
```diff
@@ -199,12 +199,12 @@ export class Eth2Gossipsub extends GossipSub {
-      item
-    );
+      return this.publishObject<GossipType.beacon_block_and_blobs_sidecar>(
+        {type: GossipType.beacon_block_and_blobs_sidecar, fork},
```
Wrong gossip type?

g11tech replied:
resolved :+1:

dapplion commented on packages/beacon-node/src/db/repositories/blobSidecarsArchive.ts:
```diff
@@ -16,15 +16,16 @@ export type BlockArchiveBatchPutBinaryItem = KeyValue<Slot, Uint8Array> & {
-  getId(value: deneb.BlobsSidecar): Slot {
-    return value.beaconBlockSlot;
+  getId(value: deneb.BlobSidecars): Uint8Array {
+    const {slot} = value[0]
+    return slot;
```
```suggestion
    return value[0].slot;
```

kevaundray commented on packages/beacon-node/src/chain/validation/blobSidecar.ts:
```diff
@@ -0,0 +1,154 @@
+  // would also be available
+
+  // [REJECT] The KZG proof is a correctly encoded compressed BLS G1 Point
+  // -- i.e. blsKeyValidate(blobs_sidecar.kzg_aggregated_proof)
```
Noting: `kzg_aggregated_proof` being used here

kevaundray commented on packages/beacon-node/src/chain/validation/blobSidecar.ts:
```diff
@@ -0,0 +1,154 @@
+  blobs: deneb.Blobs,
+  proofs: deneb.KZGProofs
+): void {
+  // assert verify_aggregate_kzg_proof(blobs, expected_kzg_commitments, kzg_aggregated_proof)
```
Noting comment re `kzg_aggregated_proof`.

(Its still a draft so I imagine you were on your way to clean these up anyways :) )

kevaundray commented on packages/beacon-node/src/chain/validation/blobSidecar.ts:
````diff
@@ -0,0 +1,154 @@
+ * ```
+ * Check that each FIELD_ELEMENT as a uint256 < BLS_MODULUS
+ */
+function blobIsValidRange(blob: deneb.Blob): boolean {
````
Noting here that whenever you make a blob proof, the cryptography code will also do this check

g11tech replied:
i am now assuming that verifyBlobKzgProof  will verify blobsidecar's  all three blob, kzg and proof as valid points as per requirement

kevaundray commented on packages/beacon-node/src/chain/validation/blobSidecar.ts:
```diff
@@ -0,0 +1,154 @@
+function blobIsValidRange(blob: deneb.Blob): boolean {
+  for (let i = 0; i < FIELD_ELEMENTS_PER_BLOB; i++) {
+    const fieldElement = blob.subarray(i * BYTES_PER_FIELD_ELEMENT, (i + 1) * BYTES_PER_FIELD_ELEMENT);
+    const fieldElementBN = bytesToBigInt(fieldElement, "be");
```
We use little endian internally in the cryptography layer, so its a bit suprising that we are using big endian to convert it to an integer here. 

Moreover, this seems like an abstraction leak and perhaps the cryptography code could expose a `isValidBlob` method to cover this

g11tech replied:
yes that would be useful, but i guess verifyBlobKzgProof will verify all three blob, kzg and proof, let me know if that assumption is not correct

kevaundray commented on packages/beacon-node/src/chain/validation/blobSidecar.ts:
```diff
@@ -0,0 +1,154 @@
+    });
+  }
+
+  validateBlobsAndProofs([blobSidecar.kzgCommitment], [blobSidecar.blob], [blobSidecar.kzgProof]);
```
@g11tech educated me on why we have BlsKeyValidate and BlobIsValidRange here and we could remove those as they are being done when we validate the proof with the blob.

(Whether you will be able to get the same amount of error information will depend on c-kzg errors)

g11tech replied:
removed explicit validations to totally rely on underlying kzg lib

dapplion commented on packages/beacon-node/src/api/impl/validator/index.ts:
```diff
@@ -272,6 +272,27 @@ export function getValidatorApi({
 
+    async getBlob(blockRoot, index) {
+      const blockRootHex = toHex(blockRoot);
+      const blob = (chain.producedBlobSidecarsCache.get(blockRootHex)?.blobSidecars ?? []).slice(index, index + 1)[0];
```
Why `.slice(index, index+1)[0]` instead of just `[index]`?

g11tech replied:
thanks for spotting, my head must be spinning

dapplion commented on packages/beacon-node/src/chain/archiver/archiveBlocks.ts:
```diff
@@ -73,28 +71,11 @@ export async function archiveBlocks(
-  // Delete expired blobs
-  // Keep only `[max(GENESIS_EPOCH, current_epoch - MIN_EPOCHS_FOR_BLOBS_SIDECARS_REQUESTS), current_epoch]`
-  if (finalizedPostDeneb) {
-    const blobsSidecarMinEpoch = currentEpoch - config.MIN_EPOCHS_FOR_BLOBS_SIDECARS_REQUESTS;
```
Where are blobs pruned now?

g11tech replied:
there is no `blobsSidecarArchive` as they are just deleted using `pruneBlobSidecarsFromHotDb` (instead of earlier doing `migrateBlobsSidecarFromHotToColdDb`) over here: https://github.com/ChainSafe/lodestar/pull/5181/files#diff-31bca7e499ec6e9d38d9ce318c6d239d35fa4f6b687213f84b75da217ac2415bR131

dapplion replied:
But without blobsSidecarArchive, do we persist blobs from finalized blocks?

g11tech replied:
no, we don't, we just prune them out of hot db using `pruneBlobSidecarsFromHotDb`

dapplion replied:
How are you gonna serve this ReqResp requests? https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md#blobsidecarsbyrange-v1 you need to serve finalized blobs

g11tech replied:
right, i mistook that even for by range requests, finalized blobs need not be served
I will re-add the functionality

dapplion commented on .github/workflows/test-sim-merge.yml:
```diff
@@ -12,8 +12,9 @@ env:
   MERGEMOCK_IMAGE: g11tech/mergemock:latest
   GETH_WITHDRAWALS_IMAGE: g11tech/geth:withdrawalsfeb8
-  ETHEREUMJS_WITHDRAWALS_IMAGE: g11tech/ethereumjs:feb8
+  ETHEREUMJS_WITHDRAWALS_IMAGE: g11tech/ethereumjs:latest
```
I know it's wip draft but is there a commit based tag to prevent spurious failures?

g11tech replied:
will add :+1:

g11tech committed:
Free the blobs

fix the types

rejig the new constants in params

add comment for cleanup

update reqresp

fix api package

rename blobs repo

commit the wip modifications

further appropriate renaming

further references update

further reference updates

continue refac

fix reqresp build

further refac

further refac

fix api

fix db interface

fix beacondb alloc

build

fix api

improve blob verificaion

correct validation call

fixes

fix the produce block/blobs flow

reduce diff

blob gossip validation

update validations

cleanup block vali

reduce diff

handle gossip of block and blob

fix test for timebeing

modify publishing flow

fix import flow

onsidecarbyrange fix and some type fixes

fix sidecars by root

prune blockinput cache

fix kzg interface

small renaming

interface rename

fix fetch blockmaybeblobs by range test

fix build lint issues for now

c-kzg version fix

FullOrBlindedBlobSidecar changes

fix tests

complete the blob publishing flow

fix test

get the single node run functional

get the gossip blob flow working

fix peer syncing using req/resp

fix sidecar by root check

refactor blobsidecars hotdb and remove archive

add blob gossip validation flow

fix topic

fix the validation condition

add blob validation and test various sync modes

fix tests

rebase fixes

enable deneb spec tests

make blobsbyroot multi block

fixes

cleanup defunt builder endpoint

archive blobs post finalization uptill the blob window

serve finalized blobs within the blob prune window

fix test

fix test

lookup in archive as well

cleanup and improvements

rebase fx

Add 4844 sim test and override the field elements per blob

update image

add blob test

add test run in package

start unknown sync and range sync

finalize the sims

change the signing flow

fix test types

fix tests

fix test

lint

update tx type and corresponding ethereumjs image

update c-kzg and use blobs bundle proof

fix test

fix test

merge getblobsbundle into getpayloadv3

update images

fix genesis config

rebase fixes

fix test

update images

fix tests

lint

fix the sidecar request count limit

fix test

lint

rebase fixes

cleanup

fix unit tests

update kzg to big endian

devnet 6 integration

fix passing setup arg

update path

fix tests

fix blobs sidecar by range response

cleanup blob import vals

reduce diff

cleanups

disable blob sim tests till spec stablizes

handle blob bytes in blockinput

fixes

twoeths commented on packages/beacon-node/src/network/processor/gossipHandlers.ts:
```diff
@@ -118,6 +139,57 @@ export function getGossipHandlers(modules: ValidatorFnsModules, options: GossipH
+    const delaySec = chain.clock.secFromSlot(slot, seenTimestampSec);
+    const recvToVal = Date.now() / 1000 - seenTimestampSec;
+
+    const {blockInput, blockInputMeta} = getBlockInput.getGossipBlockInput(config, {
```
if block hasn't come I think we should search for unknown block root here `network.searchUnknownSlotRoot({slot, root: rootHex});`

g11tech replied:
there is no gurrantee which comes first, block or blobs, unkown parent search will be triggered anyway if the next slot's block or blob comes (i will confirm the same)

g11tech replied:
may be we start search when delay > some 4000 ms?

g11tech replied:
have added a todo for this optimization to be done in the followups

twoeths commented on packages/beacon-node/src/network/processor/gossipHandlers.ts:
```diff
@@ -118,6 +139,57 @@ export function getGossipHandlers(modules: ValidatorFnsModules, options: GossipH
+      if (e instanceof BlockGossipError) {
+        // Don't trigger this yet if full block and blobs haven't arrived yet
+        if (e instanceof BlockGossipError && e.type.code === BlockErrorCode.PARENT_UNKNOWN && blockInput !== null) {
+          logger.debug("Gossip block has error", {slot, root: blockHex, code: e.type.code});
```
```suggestion
           logger.debug("Gossip blob has error", {slot, root: blockHex, code: e.type.code});
```

twoeths commented on packages/beacon-node/src/chain/blocks/types.ts:
```diff
@@ -32,26 +32,121 @@ export function blockRequiresBlobs(config: ChainForkConfig, blockSlot: Slot, clo
+      // block is available, check if all blobs have shown up
+      const {blobKzgCommitments} = (signedBlock as deneb.SignedBeaconBlock).message.body;
+      if (blobKzgCommitments.length < blockCache.blobs.size) {
+        throw Error(`Received more blobs=${blockCache.blobs.size} than commitments=${blobKzgCommitments.length}`);
```
we may want to also include block hex + slot here

g11tech replied:
done

twoeths reviewed (commented):
looks great, I dropped some minor comments

g11tech committed:
apply feedback

g11tech committed:
add todo for pre-emptive block/blobs pull

g11tech committed:
rebase fix

dapplion reviewed (approved):
Looks good!

wemeetagain commented:
:tada: This PR is included in [v1.11.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.11.0) :tada:
